tests/tracing.rs

```diff
@@ -0,0 +1,143 @@
+//! LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//! OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+//! SOFTWARE.
+
+use std::io;
+use std::sync::{Mutex, MutexGuard};
+use tracing::{Dispatch, Instrument};
+use tracing_subscriber::fmt::MakeWriter;
+use tracing_subscriber::FmtSubscriber;
+use xtra::prelude::*;
+
+#[derive(Debug)]
+pub struct MockWriter<'a> {
+    buf: &'a Mutex<Vec<u8>>,
+}
+
+impl<'a> MockWriter<'a> {
+    /// Create a new `MockWriter` that writes into the specified buffer (behind a mutex).
+    pub fn new(buf: &'a Mutex<Vec<u8>>) -> Self {
+        Self { buf }
+    }
+
+    /// Give access to the internal buffer (behind a `MutexGuard`).
+    fn buf(&self) -> io::Result<MutexGuard<'a, Vec<u8>>> {
+        // Note: The `lock` will block. This would be a problem in production code,
+        // but is fine in tests.
+        self.buf
+            .lock()
+            .map_err(|_| io::Error::from(io::ErrorKind::Other))
+    }
+}
+
+impl<'a> io::Write for MockWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // Lock target buffer
+        let mut target = self.buf()?;
+
+        // Write to stdout in order to show up in tests
+        print!("{}", String::from_utf8(buf.to_vec()).unwrap());
+
+        // Write to buffer
+        target.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.buf()?.flush()
+    }
+}
+
+impl<'a> MakeWriter<'_> for MockWriter<'a> {
+    type Writer = Self;
+
+    fn make_writer(&self) -> Self::Writer {
+        MockWriter::new(self.buf)
+    }
+}
+
+/// Return a new subscriber that writes to the specified [`MockWriter`].
+///
+/// [`MockWriter`]: struct.MockWriter.html
+pub fn get_subscriber(mock_writer: MockWriter<'static>, env_filter: &str) -> Dispatch {
+    FmtSubscriber::builder()
+        .with_env_filter(env_filter)
+        .with_writer(mock_writer)
+        .with_level(true)
+        .with_ansi(false)
+        .into()
+}
+
+pub fn logs_assert<F>(f: F)
+where
+    F: Fn(&[&str]) -> bool,
+{
+    let buf = GLOBAL_BUF.lock().unwrap();
+    let logs: Vec<&str> = std::str::from_utf8(&buf)
+        .expect("Logs contain invalid UTF8")
+        .lines()
+        .collect();
+    assert!(f(&logs))
+}
+
+lazy_static::lazy_static! {
+    /// The global log output buffer used in tests.
+    #[doc(hidden)]
+    pub static ref GLOBAL_BUF: Mutex<Vec<u8>> = Mutex::new(vec![]);
+}
+
+struct Tracer;
+
+#[async_trait]
+impl Actor for Tracer {
+    type Stop = ();
+    async fn stopped(self) {}
+}
+
+struct Hello(&'static str);
+
+#[async_trait]
+impl Handler<Hello> for Tracer {
+    type Return = ();
+
+    async fn handle(&mut self, message: Hello, _ctx: &mut Context<Self>) {
+        tracing::info!("Hi {}", message.0)
+    }
+}
+
+#[tokio::test]
+async fn assert_send_is_child_of_span() {
+    let mock_writer = MockWriter::new(&GLOBAL_BUF);
```

Comment:
Do we need to use a global buffer here? Looks trivial to use a function-local buffer which would allow us to write more of these tests that can be run in parallel, plus we can drop the lazy_static dependency.

Reply:
Good point!

Reply:
If we wrap the buffer in a new type, we can hide the details of converting the bytes to a lines iterator.

Reply:
Is there much point when this will mostly just be used through `with_logs` anyway?